File: lib/combo/decompressor.rb
require 'yaz0'
require 'fileutils'
require 'digest'
require 'combo/common'
require 'combo/dma_data'

module Combo::Decompressor
  def self.run(game)
    meta = Combo::METADATA[game]
    dst_dir = File.join(Combo::PATH_BUILD, "roms")
    stamp = File.join(dst_dir, "#{game}.stamp")
    FileUtils.mkdir_p(dst_dir)

    return if File.exists?(stamp)

    puts "Decompressing #{game}..."

    File.open(File.join(Combo::PATH_ROMS, "#{game}.z64"), "rb") do |src|
      # Check the hash of the ROMs
      hash = Digest::SHA256.hexdigest(src.read)
      valid_hashes = meta[:sha256]

      unless valid_hashes.include?(hash)
        raise "Invalid ROM hash for #{game}: #{hash}"
      end

      # Read and copy the DMA Data
      src.seek(meta[:dma_addr])
      dma = Combo::DmaData.new(src.read(meta[:dma_count] * 0x10))
      File.open(File.join(dst_dir, "#{game}_dma.bin"), "wb") do |f|
        f.write(dma.data)
      end

      File.open(File.join(dst_dir, "#{game}_decompressed.z64"), "wb") do |dst|
        # Prepare the 64 MiB rom
        dst.truncate(64 * 1024 * 1024)

        # Copy data
        dma.each do |i, e|
          next if e.dummy?
          src.seek(e.pstart)
          dst.seek(e.vstart)
          if e.compressed?
            Yaz0.decompress(src, dst)
          else
            data = src.read(e.plen)
            dst.write(data)
          end
          dma.set_decompressed(i)
        end

        # Write the new DMA data
        dst.seek(meta[:dma_addr])
        dst.write(dma.data)
      end
    end
    FileUtils.touch(stamp)
  end
end

File: lib/Decompressor.ts
import { createYaz0Stream } from 'yaz0';
import { Readable } from 'stream';
import fs from 'fs/promises';
import path from 'path';

import { DmaData } from './DmaData';
import { CONFIG, Game, GAMES, PATH_BUILD, PATH_ROMS } from './Config';

export const copyFile = async (src: Buffer, dst: Buffer, compressed: boolean) => {
  let stream = Readable.from(src);
  if (compressed) {
    stream = stream.pipe(createYaz0Stream('decompress'));
  }
  let offset = 0;
  for await (const chunk of stream) {
    chunk.copy(dst, offset);
    offset += chunk.length;
  }
  return null;
};

export const decompressGame = async (game: Game, rom: Buffer): Promise<Buffer> => {
  const conf = CONFIG[game];
  const dma = new DmaData(Buffer.from(rom.subarray(conf.dmaAddr, conf.dmaAddr + conf.dmaCount * 16)));
  const out = Buffer.alloc(64 * 1024 * 1024);
  const promises: Promise<null>[] = [];

  for (let i = 0; i < dma.count(); ++i) {
    const record = dma.read(i);
    let physEnd = record.physEnd;
    let compressed = true;
    if (physEnd === 0xffffffff) {
      continue;
    }
    if (physEnd === 0) {
      physEnd = record.physStart + (record.virtEnd - record.virtStart);
      compressed = false;
    }
    dma.write(i, { virtStart: record.virtStart, virtEnd: record.virtEnd, physStart: record.virtStart, physEnd: 0 });
    const src = rom.subarray(record.physStart, physEnd);
    const dst = out.subarray(record.virtStart, record.virtEnd);
    promises.push(copyFile(src, dst, compressed));
  }

  await Promise.all(promises);

  dma.data().copy(out, conf.dmaAddr);

  return out;
};

export const decompressGames = async () => {
  await fs.mkdir(path.resolve(PATH_BUILD, 'roms'), { recursive: true });
  for (const g of GAMES) {
    console.log("Decompressing " + g + "...");
    const rom = await fs.readFile(path.resolve(PATH_ROMS, `${g}.z64`));
    const out = await decompressGame(g, rom);
    await fs.writeFile(path.resolve(PATH_BUILD, 'roms', `${g}_decompressed.z64`), out);
  }
};
